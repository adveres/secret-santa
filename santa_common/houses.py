"""
Definitions for households
"""


class Household:
    def __init__(self, address_1: str, address_2: str, notes: str = None):
        """
        Describes a home/address.
        :param address_1:  Addr line 1
        :param address_2:  Addr line 2
        :param notes: Any special notes for this address for the delivery people to know.
        """
        self.address_1 = address_1
        self.address_2 = address_2
        self.notes = notes

    def __hash__(self):
        return hash((self.address_1, self.address_2, self.notes))

    def __eq__(self, other):
        return hash(self) == hash(other)

    def __repr__(self):
        return self.address_1 + " " + self.address_2

# TODO create households and put them here
