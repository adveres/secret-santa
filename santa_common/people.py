"""
A module for describing people.
"""
from typing import List
from .houses import Household


class Person:
    def __init__(
        self,
        first: str,
        last: str,
        email: str,
        household: Household,
        cc: List[str] = None,
    ):
        self.first = first
        self.last = last
        self.email = email
        self.household = household
        self.cc = cc

    def __hash__(self):
        return hash((self.first, self.last, self.email, self.household))

    def __eq__(self, other):
        return hash(self) == hash(other)

    def __repr__(self):
        return f"{self.first} {self.last}"


# TODO create people and put them in here
EVERYONE = ()
