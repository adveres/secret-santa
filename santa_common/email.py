"""
Helps create email messages
"""
import datetime
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import Dict

from .people import Person

YEAR = datetime.datetime.now().year


def make_email_content(giver: Person, receiver: Person) -> Dict[str, str]:
    # Create the plain-text and HTML version of your message
    text = f"""
    Hi, {giver.first}!

    You've been assigned {receiver.first} as your secret-Santa recipient.
    We're going to bring the gifts in-person, but should you need to mail them...
    Their full address is: 

    {receiver.first} ${receiver.last}
    {receiver.household.address_1}
    {receiver.household.address_2}
    
    Remember the price limit is $30!
    We can start a big email thread with peoples' wish lists separately.
    This is secret, don't spill who was assigned to you! 🙊
    If shipping, double check shipping information before you send your gift.

    Merry Christmas!
    """
    if receiver.household.notes is not None:
        text += "\n\n" + receiver.household.notes

    # ----------------- HTML -----------------
    html = f"""\
    <html>
      <body>
        <p>Hi, <b>{giver.first}</b>!<br><br>
        <h2>🎁 Who? 🎁</h2>
        You've been assigned <b>{receiver.first}</b> ({receiver.email}) as your Secret Santa recipient. <br>
        We're going to bring the gifts in-person. Should you need to mail them, their full address is: <br>
        
        <blockquote>
        <b>{receiver.first} {receiver.last}</b><br>
        <b>{receiver.household.address_1}</b><br>
        <b>{receiver.household.address_2}</b><br>
        </blockquote>
        <h2>🎁 Details 🎁</h2>
        <ul>
        <li>The price limit is <b>$30</b></li>
        <li>We can start a big email thread with peoples' wish lists separately</li>
        <li>This is secret, don't spill who was assigned to you! 🙊</li>
        <li>If shipping, double check shipping information before you send your gift.</li>
    """
    if receiver.household.notes is not None:
        html += (
            "<li><b><u>Special notes for this address:</u></b> "
            + receiver.household.notes
            + "</li>"
        )
    html += "</ul><br>🎄 Merry Christmas! 🎄</p></body></html>"

    return {"plain": text, "html": html}


def make_mime_message(
    sender_email: str, giver: Person, receiver: Person
) -> MIMEMultipart:
    message = MIMEMultipart("alternative")
    message[
        "Subject"
    ] = f"🎅 Pst {giver.first}, your {YEAR} Secret Santa pick is here! 🎅"
    message["From"] = sender_email
    message["To"] = giver.email

    # Optional CC of giver (if a kid needs to also give a gift but doesn't shop)
    if giver.cc is not None:
        message["CC"] = ",".join(
            giver.cc
        )  # Seems to work for multiple emails or single ones.

    email_content = make_email_content(giver, receiver)

    # Turn these into plain/html MIMEText objects
    part1 = MIMEText(email_content["plain"], "plain")
    part2 = MIMEText(email_content["html"], "html")

    # Add HTML/plain-text parts to MIMEMultipart message
    # The email client will try to render the last part first
    message.attach(part1)
    message.attach(part2)

    return message
