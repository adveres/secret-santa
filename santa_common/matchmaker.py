"""
Helps pair people up for secret santa
"""
import random
from typing import Tuple, List
from .people import Person
import logging

MAX_TRIES = 10


def make_pairs(all_people: Tuple[Person]) -> List[Tuple[Person, Person]]:
    pairs = []
    tries = 0

    # Basically we loop because it's not always perfect when we assign someone to a gifter. It's possible we exhaust
    # the valid receivers before we do all of the gifters because people life in the same household or something.
    # So we loop and this seems to work (almost) all of the time. At the low, low cost of CPU!
    while tries < MAX_TRIES:
        tries += 1
        try:
            pairs = _try_pair(all_people)
        except IndexError as ie:
            logging.info(
                "IndexError because we cannot always slice people into pairs evenly. Trying again... (%d)",
                tries,
            )
        else:
            break

    if not pairs:
        raise ValueError(f"After {MAX_TRIES} tries we failed to make valid pairs.")
    return pairs


def _try_pair(all_people):
    receivers = list(all_people)[:]  # copy
    pairs = []

    # Assign each giver one receiver that is not from their household
    for giver in all_people:
        # Sometimes this throws an IndexError because unluckily there is NO receiver not in the giver's household
        receiver = random.choice(
            [r for r in receivers if r.household != giver.household]
        )

        receivers.remove(receiver)
        pairs.append((giver, receiver))

    return pairs
