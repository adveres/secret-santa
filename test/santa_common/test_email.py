from santa_common.email import make_email_content

from santa_common.people import Person


def test_make_email_content(person_simple: Person, person_parent: Person, snapshot):
    res = make_email_content(person_simple, person_parent)

    print(res["plain"])
    print(res["html"])

    assert res == snapshot

    # Convenient to see the HTML in-browser sometimes
    #
    # with open("zz_test_output.html", "w") as my_file:
    #     my_file.write(res["html"])
