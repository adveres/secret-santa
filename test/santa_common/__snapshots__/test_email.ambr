# name: test_make_email_content
  dict({
    'html': '''
          <html>
            <body>
              <p>Hi, <b>Simple</b>!<br><br>
              <h2>游꾸 Who? 游꾸</h2>
              You've been assigned <b>Parent</b> (fakeEmail@fakeEmailsRus.com) as your Secret Santa recipient. <br>
              We're going to bring the gifts in-person. Should you need to mail them, their full address is: <br>
              
              <blockquote>
              <b>Parent Person</b><br>
              <b>1111 Robot Lane</b><br>
              <b>Epcot, FL 99999</b><br>
              </blockquote>
              <h2>游꾸 Details 游꾸</h2>
              <ul>
              <li>The price limit is <b>$30</b></li>
              <li>We can start a big email thread with peoples' wish lists separately</li>
              <li>This is secret, don't spill who was assigned to you! 游뗽</li>
              <li>If shipping, double check shipping information before you send your gift.</li>
          <li><b><u>Special notes for this address:</u></b> Some Amazon companies won't ship to a PO box, so the gifter may need to use the Amazon store that's downtown for pickups.</li></ul><br>游꾻 Merry Christmas! 游꾻</p></body></html>
    ''',
    'plain': '''
      
          Hi, Simple!
      
          You've been assigned Parent as your secret-Santa recipient.
          We're going to bring the gifts in-person, but should you need to mail them...
          Their full address is: 
      
          Parent $Person
          1111 Robot Lane
          Epcot, FL 99999
          
          Remember the price limit is $30!
          We can start a big email thread with peoples' wish lists separately.
          This is secret, don't spill who was assigned to you! 游뗽
          If shipping, double check shipping information before you send your gift.
      
          Merry Christmas!
          
      
      Some Amazon companies won't ship to a PO box, so the gifter may need to use the Amazon store that's downtown for pickups.
    ''',
  })
# ---
